src/main/java/hudson/tasks/junit/CaseResult.java

```diff
@@ -256,10 +262,25 @@ private static String getSkippedMessage(Element testCase) {
         return message;
     }
 
-    public String getDisplayName() {
+    public String getTransformedTestName() {
         return TestNameTransformer.getTransformedName(testName);
     }
 
+    public String getDisplayName() {
+        // Only prepend the enclosing flow node names if there are any and the run this is in has multiple blocks directly containing
+        // test results.
+        if (!getEnclosingFlowNodeNames().isEmpty()) {
```

Comment:
I think it is - given that the desired functionality is "tell me what stage/branch/etc this test was reported in", we either need to record the stage/branch/etc information (i.e., the enclosing flow nodes) in the test result, at which point we've got it baked in that one time we write to it and never have to query again, or we need to take the step flow node where the test was recorded and, every time we want to report on tests, determine what stage/branch/etc that step was in.

The two primary "use cases" I've come across for this so far are:
* Getting the "fully qualified" name (including stage/branch) for every test in a run for display purposes, i.e., the Blue Ocean or classic Jenkins test result screens
* Getting the list of tests recorded somewhere within a stage/branch, i.e., to have a more granular version of the existing test result screens scoped to a specific stage/branch

The former is the one I've currently implemented here and in Blue Ocean, fwiw.

Both of those use cases are dramatically easier if enclosing flow nodes are baked into the test result. As demonstrated here with `CaseResult#getDisplayName()`, the only complexity added to the display logic here in `junit` is to check if we need to prepend the stage/branch name(s) at all (i.e., if there are multiple enclosing stages/branches amongst all the test results for this run) - and even that is working entirely from data under the `TestResultAction`. To do this without recording the stage/branch info at runtime, each time we wanted to display the results, we'd have to load the `FlowExecution`, find the `FlowNode` where the test result was recorded, find its enclosing stage/branch, etc - yeah, we could cache that (and would have to in practice, I think, since the alternative would be doing that query path for every single test over and over when we need to check if we should display the stage/branch names in the first place) but it still exposes us to the same costs as doing it at runtime, with the strong possibility of doing it all more than once (i.e., when the cache expires, when the master restarts, etc).

And the second use case...well, we'd have to determine the flow nodes contained within a given stage/branch from the `FlowExecution` and then find all test results recorded in one of those contained flow nodes. And that doesn't seem to be an easy thing to do without some chunk visitor fun, to say the least.

I'd definitely like a more optimized way to say "get me the names, in some predictable order, of any stage or parallel branch containing my current flow node" at runtime.